adaptive/learner/average_learner1D.py

```diff
@@ -0,0 +1,435 @@
+        if x_right is not None:
+            self._distances[x] = hypot((x_right - x), (self.data[x_right] - y))
+
+    def _update_losses_resampling(self, x, real=True):
+        """Update all losses that depend on x, whenever the new point is a re-sampled point."""
+        # (x_left, x_right) are the "real" neighbors of 'x'.
+        x_left, x_right = self._find_neighbors(x, self.neighbors)
+        # (a, b) are the neighbors of the combined interpolated
+        # and "real" intervals.
+        a, b = self._find_neighbors(x, self.neighbors_combined)
+
+        if real:
+            for ival in _get_intervals(x, self.neighbors, self.nth_neighbors):
+                self._update_interpolated_loss_in_interval(*ival)
+        elif x_left is not None and x_right is not None:
+            # 'x' happens to be in between two real points,
+            # so we can interpolate the losses.
+            dx = x_right - x_left
+            loss = self.losses[x_left, x_right]
+            self.losses_combined[a, x] = (x - a) * loss / dx
+            self.losses_combined[x, b] = (b - x) * loss / dx
+
+        # (no real point left of x) or (no real point right of a)
+        left_loss_is_unknown = (x_left is None) or (not real and x_right is None)
+        if (a is not None) and left_loss_is_unknown:
+            self.losses_combined[a, x] = float("inf")
+
+        # (no real point right of x) or (no real point left of b)
+        right_loss_is_unknown = (x_right is None) or (not real and x_left is None)
+        if (b is not None) and right_loss_is_unknown:
+            self.losses_combined[x, b] = float("inf")
+
+    def _calc_error_in_mean(self, ys, y_avg, n):
+        variance_in_mean = sum((y - y_avg) ** 2 for y in ys) / (n - 1)
+        t_student = scipy.stats.t.ppf(1 - self.alpha, df=n - 1)
+        return t_student * (variance_in_mean / n) ** 0.5
+
+    def tell_many(self, xs, ys):
+        """Tell the learner about some values.
+
+        Parameters
+        ----------
+        xs : Iterable of values from the function domain
+        ys : Iterable of lists of values (several data samples)
+             or scalars from the function image (mean values that
+             are directly included in self.data; the number of samples is then
+             assumed to be 1 and the error 0 for all data points; these points
+             will not be included in _rescaled_error_in_mean and therefore
+             will not be resampled)
+
+        Examples
+        ----------
+        If xs = [0,1] and ys = [[1.2,1.4],[2]], the y-values will be included in
+        self._data_samples.
+        If xs = [0,1] and ys = [1.2,2], the y-values will be directly included
+        in self.data, and the previous self.data[0] and self.data[1] will be erased.
+        """
+        # Check that all x are within the bounds
+        assert np.prod([x>=self.bounds[0] and x<=self.bounds[1] for x in xs]), 'x value out of bounds: remove x or enlarge the bounds of the learner'
```

Comment:
This may not be an `assert`: assertions are for internal sanity checks. Instead we should `raise ValueError` here.

Reply:
Fixed [here](https://github.com/python-adaptive/adaptive/pull/283/commits/a8c86ccae75e0f5d4da721bd1792f8c9a10c95e8)!

Reply:
In other learners one is able to have numbers outside of the domain, the only thing that should happen is that it doesn't suggest points outside of the domain.